### PR TITLE
SWC-2248 (SWC 2272)

### PR DESCRIPTION
**Pull request by kimyen**

### CODE REVIEW

File: src/main/java/org/sagebionetworks/web/client/widget/entity/download/Uploader.java [hunk 1]
```diff
@@ -205,6 +206,7 @@ public void updateS3UploadBannerView(String banner) {
 	
 	public void queryForUploadDestination() {
 		enableMultipleFileUploads();
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/widget/entity/download/Uploader.java:208]: Please add a few tests around this method, to verify that the storage location id has the expected value when the first UploadDestination is of different types (if is empty or if no parent then storage location should be null, if S3 or External then it should be whatever was set in the UploadDestination)

File: src/test/java/org/sagebionetworks/web/unitclient/widget/entity/download/UploaderTest.java [hunk 2]
```diff
@@ -331,7 +332,7 @@ public void testUploadToExternalInvalid() {
 		destinations.add(d);
 		AsyncMockStubber.callSuccessWith(destinations).when(synapseClient).getUploadDestinations(anyString(), any(AsyncCallback.class));
 		uploader.queryForUploadDestination();
-		
+		assertNull(uploader.getStorageLocationId());
```
Comment (kimyen) [inline src/test/java/org/sagebionetworks/web/unitclient/widget/entity/download/UploaderTest.java:335]: case invalid External Upload Destination

File: src/test/java/org/sagebionetworks/web/unitclient/widget/entity/download/UploaderTest.java [hunk 3]
```diff
@@ -342,6 +343,7 @@ public void testInvalidUploadDestination() {
 		destinations.add(mock(UploadDestination.class));
 		AsyncMockStubber.callSuccessWith(destinations).when(synapseClient).getUploadDestinations(anyString(), any(AsyncCallback.class));
 		uploader.queryForUploadDestination();
+		assertNull(uploader.getStorageLocationId());
```
Comment (kimyen) [inline src/test/java/org/sagebionetworks/web/unitclient/widget/entity/download/UploaderTest.java:346]: case invalid Upload Destination

File: src/test/java/org/sagebionetworks/web/unitclient/widget/entity/download/UploaderTest.java [hunk 4]
```diff
@@ -355,12 +357,14 @@ public void testUploadToExternal() {
...
 		d.setUploadType(UploadType.SFTP);
 		d.setUrl(url);
+		d.setStorageLocationId(storageLocationId);
 		List<UploadDestination> destinations = new ArrayList<UploadDestination>();
 		destinations.add(d);
 		AsyncMockStubber.callSuccessWith(destinations).when(synapseClient).getUploadDestinations(anyString(), any(AsyncCallback.class));
 		AsyncMockStubber.callSuccessWith("ok.net").when(synapseClient).getHost(anyString(), any(AsyncCallback.class));
 		
 		uploader.queryForUploadDestination();
+		assertEquals(uploader.getStorageLocationId(), storageLocationId);
```
Comment (kimyen) [inline src/test/java/org/sagebionetworks/web/unitclient/widget/entity/download/UploaderTest.java:367]: case External Upload Destination

File: src/test/java/org/sagebionetworks/web/unitclient/widget/entity/download/UploaderTest.java [hunk 5]
```diff
@@ -376,8 +380,18 @@ public void testUploadToExternal() {
...
+	public void testQueryForUploadDestinationsWithUploadToS3() {
+		S3UploadDestination d = new S3UploadDestination();
+		d.setUploadType(UploadType.S3);
+		d.setStorageLocationId(storageLocationId);
+		List<UploadDestination> destinations = new ArrayList<UploadDestination>();
+		destinations.add(d);
+		AsyncMockStubber.callSuccessWith(destinations).when(synapseClient).getUploadDestinations(anyString(), any(AsyncCallback.class));
+		uploader.queryForUploadDestination();
+		assertEquals(uploader.getStorageLocationId(), storageLocationId);
+	}
```
Comment (kimyen) [inline src/test/java/org/sagebionetworks/web/unitclient/widget/entity/download/UploaderTest.java:394]: case S3 Upload Destination

File: src/test/java/org/sagebionetworks/web/unitclient/widget/entity/download/UploaderTest.java [hunk 6]
```diff
@@ -389,12 +403,18 @@ public void testQueryForUploadDestinationsWithoutParentEntityId() {
 		
 		Mockito.reset(synapseClient);
 		uploader.asWidget(fileEntity);
-		
+		assertNull(uploader.getStorageLocationId());
```
Comment (kimyen) [inline src/test/java/org/sagebionetworks/web/unitclient/widget/entity/download/UploaderTest.java:406]: case null Parent Entity Id

File: src/test/java/org/sagebionetworks/web/unitclient/widget/entity/download/UploaderTest.java [hunk 7]
```diff
@@ -389,12 +403,18 @@ public void testQueryForUploadDestinationsWithoutParentEntityId() {
...
 		assertEquals(entityId, stringCaptor.getValue());
 	}
-	
+
+	@Test
+	public void testQueryForUploadDestinationsWithNullEntity() {
+		Mockito.reset(synapseClient);
+		uploader.asWidget((FileEntity)null);
+		assertNull(uploader.getStorageLocationId());
+	}
```
Comment (kimyen) [inline src/test/java/org/sagebionetworks/web/unitclient/widget/entity/download/UploaderTest.java:417]: case null Entity

File: src/main/java/org/sagebionetworks/web/client/widget/upload/MultipartUploaderImpl.java [hunk 8]
```diff
@@ -379,7 +381,7 @@ public void uploadSelectedFile(String fileInputId,ProgressingFileUploadHandler h
 		}
 		int index = 0;
 		String fileName = names[0];
-		uploadFile(fileName, fileInputId, index, handler);
+		uploadFile(fileName, fileInputId, index, handler, null);
```
Comment (kimyen) [inline src/main/java/org/sagebionetworks/web/client/widget/upload/MultipartUploaderImpl.java:384]: this is not for file upload in File/ Folder
make sure there are tests for it.

**Pull request merged**